pkg/chartsync/git.go

```diff
@@ -0,0 +1,337 @@
+			if err != nil {
+				continue // this should never happen
+			}
+			// Schedule release sync by adding it to the queue.
+			c.releaseQueue.AddRateLimited(cacheKey)
+		}
+	}
+}
+
+func (c *GitChartSync) get(hr *v1.HelmRelease) (sourceRef, bool) {
+	c.sourcesMu.RLock()
+	defer c.sourcesMu.RUnlock()
+	if s, ok := c.sources[hr.ResourceID().String()]; ok && s.forHelmRelease(hr) {
+		return s, ok
+	}
+	return sourceRef{}, false
+}
+
+func (c *GitChartSync) store(hr *v1.HelmRelease, s sourceRef) {
+	c.sourcesMu.Lock()
+	c.sources[hr.ResourceID().String()] = s
+	c.sourcesMu.Unlock()
+}
+
+// sync synchronizes the record we have for the given `v1.HelmRelease`
+// with the given mirror. It always updates the HEAD record in the
+// `sourceRef`, but only returns `true` if the update was relevant for
+// the release (e.g. a change in git the chart source path, or a new
+// record). In case of failure it returns an error.
+func (c *GitChartSync) sync(hr *v1.HelmRelease, mirrorName string, repo *git.Repo) (sourceRef, bool, error) {
+	source := hr.Spec.GitChartSource
+	if source == nil {
+		return sourceRef{}, false, nil
+	}
+
+	if status, err := repo.Status(); status != git.RepoReady {
+		return sourceRef{}, false, ChartNotReadyError{err}
+	}
+
+	var changed bool
+	s, ok := c.get(hr)
+	if !ok {
+		s = sourceRef{mirror: mirrorName, remote: source.GitURL, ref: source.RefOrDefault()}
+		changed = true
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.config.GitTimeout)
+	head, err := repo.Revision(ctx, s.ref)
+	cancel()
+	if err != nil {
+		return sourceRef{}, false, ChartUnavailableError{err}
+	}
+
+	if !changed {
+		// If the head still equals to what is in our books, there are no changes.
+		if s.head == head {
+			return s, false, nil
+		}
+
+		// Check if the mirror has seen commits in paths we are interested in for
+		// this release.
+		ctx, cancel = context.WithTimeout(context.Background(), c.config.GitTimeout)
+		commits, err := repo.CommitsBetween(ctx, s.head, head, source.Path)
+		cancel()
+		if err != nil {
+			return sourceRef{}, false, ChartUnavailableError{err}
+		}
+		changed = len(commits) > 0
+	}
+
+	s.head = head
+	c.store(hr, s)
+	return s, changed, nil
+}
+
+// maybeMirror requests a new mirror for the given remote. The return value
+// indicates whether the repo was already present (`true` if so,
+// `false` otherwise).
+func (c *GitChartSync) maybeMirror(mirrorName string, remote string) bool {
+	ok := c.mirrors.Mirror(
+		mirrorName, git.Remote{URL: remote}, git.Timeout(c.config.GitTimeout),
+		git.PollInterval(c.config.GitPollInterval), git.ReadOnly)
+	if !ok {
+		c.logger.Log("info", "started mirroring new remote", "remote", remote, "mirror", mirrorName)
+	}
+	return ok
+}
+
+// helmReleasesForMirror returns a slice of `HelmRelease`s that make
+// use of the given mirror.
+func (c *GitChartSync) helmReleasesForMirror(mirror string) ([]*v1.HelmRelease, error) {
+	hrs, err := c.lister.List(labels.Everything())
```

Comment:
Isn't it a bit overkill to list all the helm releases? I don't know the operator well, but I presume we use a `HelmRelease` informer, which we can use cache the mirrors and keep the updated.

Reply:
We could use the same cache to make sure that `c.sources` are up to date, allowing to garbage collect mirrors and (maybe) allowing to remove the consistency check in `get()`

Reply:
This is a `HelmReleaseLister`, which is nothing more than a named `cache.Indexer`, we basically loop over in-memory `HelmRelease` resources at close-to-zero cost.

Reply:
(I can also give you some historical insights and tell you that before my previous refactor round, we used the client for this type of operations 🤫).

Reply:
> we basically loop over in-memory HelmRelease resources at close-to-zero cost.

Ah, I see, I thought it involved calling the client 👍